### PR TITLE
[Testing] Understanding FileCache & conducting related experiments

### PR DESCRIPTION
**Pull request by honest-mule**
## Description

This is not an actual PR. I'm looking to understand the inner-workings of FileCache and how differently it will behave when its code-sections are changed as per new assumptions or knowledge.
I need access to experimental android builds that Jenkins will help build via this PR.

@thexai I request you to please add appropriate flags to this PR so that Jenkins picks it up.

## Motivation and context

Understanding the code of FileCache.
I can currently build on Windows & Linux, but churning out test-builds for Android devices is out of my reach as of now.

## How has this been tested?

NA

## What is the effect on users?

None

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [x] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by thexai** [comment]
> I request you to please add appropriate flags to this PR so that Jenkins picks it up.

Jenkins always builds all PR's by default (even without tags)

File: xbmc/filesystem/FileCache.cpp [hunk 1]
```diff
@@ -123,6 +123,7 @@ bool CFileCache::Open(const CURL& url)
   m_chunkSize = CFile::DetermineChunkSize(
       m_source.GetChunkSize(),
       CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_cacheChunkSize);
+  m_chunkSize = (m_chunkSize / 4096) * 4096;
```
Comment (fritsch) [inline xbmc/filesystem/FileCache.cpp:126]: Other way round. To care for rounding. First multiply afterwards divide.

**Comment by arnova** [comment]
FYI: I have a pretty good understanding of the code in CFileCache since most of it was either refactored or fixed by me. The big unknown (and problematic code) is the level/cache in handling in VP related to CFileCache...

**Comment by honest-mule** [comment]
Thanks for your input, @arnova.

I've recognized a weird situation, on android based firestick 4K max, Kodi is only reading data from the internet at around 100Mbps max even when a connection of 300Mbps is provided to it. Meaning even if the fire OS and/or analiti speed test is showing a rate of >250Mbps, Kodi will only utilize around 100Mbps nomatter what.

Builds tested: Stable, Nightly, Kodinerds.

What section of the code do you think might be causing this abnormality?

**Comment by thexai** [comment]
Kodi does not try (and should not try) to read a stream at the highest possible speed. Kodi must read a stream at a speed proportional (just slightly higher) to the bitrate of the stream and accumulate the difference (read - spent) in the cache, until cache is filled.

https://forum.kodi.tv/showthread.php?tid=371292&pid=3166613#pid3166613

There is a fundamental concept and that is that Kodi does NOT read an 80 Mbit/s stream at the same speed as another 0.8 Mbit/s stream.

If you want to know the limit speed you should use an imaginary test stream of 100000000 Mbit/s.

**Comment by honest-mule** [comment]
@arnova 
>Did you also try different protocols or using the disk-based cache ?
I've tested just memory based cache as of now thinking memory will always be faster than persistent storage method.

>I assume you see this behavior while it's filling up from scratch?
Always, regardless of the fact if the video is playing, halted or paused.

>And did you also trying setting the source chunk-size to something bigger?
Yes, all result in the same scenario.

>or the thread simply doesn't get enough time to do its thing
We've to be able to cross this off from the possible causes. Do you recommend (just for testing purposes) a multi-threaded read/write approach in the `FileCache->Process()` function?

-----

@thexai I agree with your sentiment, however, you should note that I'm not testing low-bitrate content. I'm doing testing with UHD BD-remuxes. Kodi should be able stream them at the required speed, and 100Mbps is nowhere near playable speed for such video files. Also, the tests do not end at trying to play the starting 2-3 minutes of the movie. I'm testing all the sections of a given movie because VBR content has peaks and troughs.

**Comment by arnova** [comment]
> I've tried HTTP `.strm` files, HTTP web-directory and WEBDAV folder. As of now I've tested just memory based cache, thinking memory will always be faster than persistent storage method.
 
Yes, that's what to be expected but you should never rule out the possibility it isn't for some reason.

 
> We've to be able to cross this off from the possible causes. Do you recommend (just for testing purposes) a multi-threaded read/write approach in the `FileCache->Process()` function?

Yes you could give that a try. Might as well be that for some reason we spend too much time in the CFile::Read()-call...

**Comment by thexai** [comment]
I can not replicate your bandwidth limitation issues, then is not Kodi but somewhat in your setup or hardware.

Used Jellyfish Video Bitrate Test Files: 
https://www.larmoire.info/jellyfish/

Even extreme 400 Mbps case plays fine and buffer fills forward with peak bitrate of 800 - 900 Mbps (measured at network card)

image

I mean Kodi (current master branch) can read files at 800 - 900 Mbps from remote NAS using NFS (also tested SMB with similar results). This with default readfactor 4.0 and only changed memorysize to 1024 MB to test read rates > video bit rate.

Note that not exist any commercial UHD Blu-Ray with these high bitrates, maximum is probably 120 - 140 Mbps and usually 80 - 90 Mbps

**Comment by thexai** [comment]
This is limiting readfactor to 1.5 video plays fine but cache not fills much:


https://github.com/xbmc/xbmc/assets/58434170/067f27d2-2f38-4098-b901-4350c1864ad7

**Comment by thexai** [comment]
This is 250 Mbps video with 10.0 readfactor


https://github.com/xbmc/xbmc/assets/58434170/e1fc3183-ff71-4995-b239-059d440698f7

**Comment by honest-mule** [comment]
>ps. Are you absolutely sure the speed isn't limited by the hardware somehow?

@arnova If firestick 4k max is only capable of so much processing, then it would be right to say that Kodi processing is being limited by the hardware.

------

@thexai On PC/Laptop, my tests give me the same results as you've posted here. Kodi grinds through large streams while filling up the cache at a very good rate.

In the earlier post, I specifically mentioned **android based firestick 4k max**. I would love it if someone can test Kodi on similar android-centric hardware and beat my findings that Kodi isn't using up more than around 100Mbps for internet video streaming.

**Comment by thexai** [comment]
On Shield 2019 h264 250 Mbps file plays fine with +50 MB forward cache and totally smooth

**Comment by thexai** [comment]
>  I specifically mentioned android based firestick 4k max.

Has firestick 4K max Gbps ethernet?

**Comment by honest-mule** [comment]
Wow, the threaded source reads (contrary to threaded cache writes) have made a superior difference on the firestick device.

Using a single detached thread for reading has made the hardware read at an astonishing rate (for its kind) of ~8.5MB/s.

World War Z UHD remux is playing smoother now, cache fills healthily during low-bitrate scenes. Of course, this bitrate isn't enough for the movie to play without interruptions but it seems like a marvelous jump from before when the max rate was reaching just ~7MB/s.

This leads me to believe the reading from the source should be done using at least 2 threads (at least for high-bitrate content on these low-powered android devices).

@arnova Would you be so kind to review the last commit? Seeking doesn't work at all since I've paid little attention towards that aspect and more towards how to utilize more bandwidth.

Also, what do you say about utilizing multiple threads to read from source?

**Comment by fritsch** [comment]
What happens if you double the chunksize instead? Change everything back and go to chunksize implementation of the File Implementation you test and set that to 1M or 2M or something?

I have the feeling that your second reader just workarounds inefficient reads.

**Comment by arnova** [comment]
Yes, probably reading bigger chunks will also workaround the issue. Note that it depends on the filesystem used since not every filesystem supports true chunked reading afaik. Like I previously said: probably the CFile::Read()-calls are relatively slow causing overhead especially with high-bitrate content

**Comment by honest-mule** [comment]
>What happens if you double the chunksize instead? Change everything back and go to chunksize implementation of the File Implementation you test and set that to 1M or 2M or something?

I've already tried this approach before I did any modifications to the code and it seemed to have little to no effect. I've tried again now with the read/write threaded test code + different chunksize method and I don't see any noticeable difference.

**Comment by fritsch** [comment]
Thanks for checking. That might maybe also mean, that the second thread workarounds a "blocking" when the fetched data is read / filled into the thread. So the second one is hanging on the mutex right away. With a performance analyzer (perf) you might see that.

Actually maybe even the CPU consumption can tell that now the second thread is "hard core" polling on the mutex ... to enqueue its data.

What do you see on that level?

**Comment by honest-mule** [comment]
There's no easy way to check for CPU usage levels on this device. On their developer pages they've screenshots showing CPU usage levels in system x-ray menu, but I think they've either removed that functionality on 4k max or not provided that in the first place, since it's completely missing in the version of the system x-ray that my device produces.

Moreover, when looking in the device specifications, I found this info:
image

Perhaps this is a clue which means their device is only capable of maximum 100Mbps video playback?

Link: https://developer.amazon.com/docs/fire-tv/device-specifications-fire-tv-streaming-media-player.html?v=ftvstick4kmax

**Comment by fritsch** [comment]
What about: 
adb shell
top

Works on my amazon device.

**Comment by fritsch** [comment]
For your specsheet: 100 Mbit/s - they don't say you get it from LAN / WLAN, just that the decoder can theoretically throughput 100 Mbit/s AV1 bitrate. Those benchmarks are usually done M2M.

**Comment by honest-mule** [comment]
Thanks for the adb command hint! Here's a screenshot of that command:

image

**Comment by fritsch** [comment]
Thread 15 is maxed out, it seems. So CPU load at around 100%? Might be visible from the debug log which is thread 15.

**Comment by honest-mule** [comment]
The load stays around 70-85% using kodinerds latest build:

image

**Comment by fritsch** [comment]
So then - you won some more bandwidth by pushing the CPU even harder, right?

Also take great care, that we see a userspace thread causing this load, not a kernel or driver or interrupt load. I could bet that when you do iperf3 with the same speed, that you don't even get barely that amount of CPU load. I think you are pushing into the wrong direction namely just "doubleing" an inefficient CPU bound memcpy implementation even further.

**Comment by honest-mule** [comment]
Yes, your analysis on point.

>Also take great care, that we see a userspace thread causing this load
Given that it's just testing that we're doing, this hardly matters because a release build is never supposed behave in this manner.

May I know which version of firestick you have? It would be nice if you can replicate this scenario at your end and see if >100Mbps wifi streaming (UHD BD-remuxes) is working as expected for you or not.

**Comment by fritsch** [comment]
I use a Cube 3rd Gen with a USB2.0 driven Gigabit dapter, which in reality (due to usb2.0 attached only) maxes out at 320 Mbit/s roughly.

**Comment by honest-mule** [comment]
Can you test playing a hefty UHD Remux on that either over WiFi6E and/or ethernet to see if it's playing smoothly or not?

You can also try thexai's link (https://www.larmoire.info/jellyfish/) to find high bitrate video samples to play. 

Would be very interesting to see which high-bitrate file is the limit for it since Amazon claims that it's processor is 2x-3x faster than of firestick 4k max.

**Comment by fritsch** [comment]
I have some raw 4K blurays ripped. They work okayish.

For jellyfish:
250 Mbit/s works like a charm
300 Mbit/s works like a charm

advancedsettings:
``<advancedsettings>
  <cache>
    <buffermode>1</buffermode>
    <memorysize>134217728</memorysize>
    <readfactor>1.5</readfactor>
  </cache>

image

  
  
</advancedsettings>

**Comment by honest-mule** [comment]
This should conclude the limits of these devices. In my opinion Kodi wiki should be updated. @thexai, what do you say about that? If I had known that firestick 4K max can't smoothly play all 4K BDremuxes hardware-wise, it would've impacted my purchase decision.

On the other hand, as per @fritsch's experience, Cube 3rd Gen on ethernet is playing up to 300mbps videos without hitches.

I propose a new wiki page dedicated just for TV boxes specifying the max bitrate file and their respective codecs, that Kodi can play on them without pauses. This would greatly help Kodi-users in swiftly summarizing which android TV box will fit their needs the best.

**Comment by fritsch** [comment]
Plugin a usb Disk first or copy the file to your internal storage. Or try with VLC as well?

**Comment by fritsch** [comment]
I will test on the 4K Max tomorrow as well - I have that also somewhere around. No Wifi 6E but Wifi5 AC with perfect reception. Will report back my findings.

**Comment by honest-mule** [comment]
Even though I have not tested that yet, I am pretty sure video files up to 300 Mbps will play smoothly via USB, but it's of little benefit since it chews away all the slack that a network drive provides.

Since my android TV plays UHD BDremuxes via USB without issues, more than likely firesticks can do that too, given they have a very similar hardware.

Firestick 4K max supports external USB sticks via OTG mechanism, but they have to be formatted in FAT32, which limits video file sizes of 4GB max. Add to this the fact that it only comes with 8GB of internal storage, of which >5GB is used by the system, which means you can only either have a thumbnail/artwork-complimented Kodi library saved on it or other streaming apps. In my case, I needed both so I had to use a sort of not-so-known mechanism to expand the storage of the device, which involved formatting the external USB via FireOS to make the external USB act as part of the system storage so that Kodi and its data can be moved to that external USB and be used to launce Kodi that way without eating up the internal storage. Given that my USB-OTG solution has only a single USB port, it means I can either run Kodi from external USB or connect a different USB having high-bitrate UHD video files divided in 4GB parts or max 4GB -- not possible at the same time. Haha.

>Or try with VLC as well?
Yes, tried it. Can't play high-bitrate UHD video files over the network without significant halts.

**Comment by fritsch** [comment]
Okay. What you describe basically means, that you bought the wrong device for your usecases - at least out of the box. In some days a new Firetv Stick is released, maybe that's something for you then. 

Still if you find another way of sharing the data, something similar to os mounts or something it would be worth a try - I don't want to suggest to try the rooted ROMs available somewhere with nfs support, it would be technically interesting, to see if a more efficient file sharing would help.

**Comment by honest-mule** [comment]
@fritsch With regards to your input I purchased the Fire TV Cube recently and indeed all the UHD blurays are playing smoothly now!

Thank you very much for your insight.

**Comment by jp12git** [comment]
@honest-mule What settings are you using on Fire TV Cube 3? I'm not able to play high bitrate files on my Cube 3 as shared here  

https://github.com/xbmc/xbmc/pull/24019#issuecomment-1786166614

**Comment by honest-mule** [comment]
@jp12git Nothing special. There are little to no options in the Fire OS to configure/customize how the device interacts with your WiFi router.

To ensure the maximum speeds possible via WiFi, your best bet would be to connect the fire tv device to a 5Ghz network and your router should be in the same room.

Furthermore you can try Kodinerds' Kodi builds for FireTV OS which I use personally. All the UHD Remuxes that I've thrown at it till now have played smoothly through and through.

Also, may I ask which specific UHD bluray is stuttering on your device?

**Comment by jp12git** [comment]
@honest-mule Interesting, I haven't used those builds before. What's different about these builds? 

I'm trying to play the samples from your link above - https://www.larmoire.info/jellyfish/

**Pull request closed without merging**